File: flink-table/flink-table-planner/src/test/scala/org/apache/flink/table/runtime/harness/BaseTwoInputStreamOperatorWithStateRetentionTest.scala
/*
 * Licensed to the Apache Software Foundation (ASF) under one
 * or more contributor license agreements.  See the NOTICE file
 * distributed with this work for additional information
 * regarding copyright ownership.  The ASF licenses this file
 * to you under the Apache License, Version 2.0 (the
 * "License"); you may not use this file except in compliance
 * with the License.  You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package org.apache.flink.table.runtime.harness

import java.lang.{Long => JLong}

import org.apache.flink.api.common.time.Time
import org.apache.flink.api.common.typeinfo.BasicTypeInfo
import org.apache.flink.runtime.state.VoidNamespace
import org.apache.flink.streaming.api.operators.InternalTimer
import org.apache.flink.streaming.runtime.streamrecord.StreamRecord
import org.apache.flink.streaming.util.KeyedTwoInputStreamOperatorTestHarness
import org.apache.flink.table.api.TableConfig
import org.apache.flink.table.runtime.harness.HarnessTestBase.{TestTableConfig, TupleRowKeySelector}
import org.apache.flink.table.runtime.join.BaseTwoInputStreamOperatorWithStateRetention
import org.apache.flink.table.runtime.types.CRow
import org.hamcrest.{Description, TypeSafeMatcher}
import org.junit.{After, Before, Test}
import org.hamcrest.MatcherAssert.assertThat

import scala.collection.mutable
import scala.collection.mutable.ArrayBuffer

/**
  * Tests for the [[BaseTwoInputStreamOperatorWithStateRetention]].
  */
class BaseTwoInputStreamOperatorWithStateRetentionTest extends HarnessTestBase {

  private val recordAForFirstKey = new StreamRecord(CRow(1L: JLong, "hello"))
  private val recordBForFirstKey = new StreamRecord(CRow(1L: JLong, "world"))

  private val config = new TestTableConfig
  config.setIdleStateRetentionTime(Time.milliseconds(2), Time.milliseconds(4))

  private var operatorUnderTest: StubOperatorWithStateTTL = _

  private var testHarness: KeyedTwoInputStreamOperatorTestHarness[JLong, CRow, CRow, CRow] = _

  @Before
  def createTestHarness(): Unit = {
    operatorUnderTest = new StubOperatorWithStateTTL(
      config.getMinIdleStateRetentionTime,
      config.getMaxIdleStateRetentionTime)
    testHarness = createTestHarness(operatorUnderTest)
    testHarness.open()
  }

  @After
  def closeTestHarness(): Unit = {
    testHarness.close()
  }

  @Test
  def normalScenarioWorks(): Unit = {
    testHarness.setProcessingTime(1L)
    testHarness.processElement1(recordAForFirstKey)

    testHarness.setProcessingTime(10L)

    assertThat(operatorUnderTest, hasFiredCleanUpTimersForTimestamps(5L))
  }

  @Test
  def whenCurrentTimePlusMinRetentionSmallerThanCurrentCleanupTimeNoNewTimerRegistered(): Unit = {
    testHarness.setProcessingTime(1L)
    testHarness.processElement1(recordAForFirstKey)

    testHarness.setProcessingTime(2L)
    testHarness.processElement1(recordBForFirstKey)

    testHarness.setProcessingTime(20L)

    assertThat(operatorUnderTest, hasFiredCleanUpTimersForTimestamps(5L))
  }

  @Test
  def whenCurrentTimePlusMinRetentionLargerThanCurrentCleanupTimeTimerIsUpdated(): Unit = {
    testHarness.setProcessingTime(1L)
    testHarness.processElement1(recordAForFirstKey)

    testHarness.setProcessingTime(4L)
    testHarness.processElement1(recordBForFirstKey)

    testHarness.setProcessingTime(20L)

    assertThat(operatorUnderTest, hasFiredCleanUpTimersForTimestamps(8L))
  }

  @Test
  def otherSideToSameKeyStateAlsoUpdatesCleanupTimer(): Unit = {
    testHarness.setProcessingTime(1L)
    testHarness.processElement1(recordAForFirstKey)

    testHarness.setProcessingTime(4L)
    testHarness.processElement2(recordBForFirstKey)

    testHarness.setProcessingTime(20L)

    assertThat(operatorUnderTest, hasFiredCleanUpTimersForTimestamps(8L))
  }

  // -------------------------------- Test Utilities --------------------------------

  private def createTestHarness(operator: BaseTwoInputStreamOperatorWithStateRetention) = {
    new KeyedTwoInputStreamOperatorTestHarness[JLong, CRow, CRow, CRow](
      operator,
      new TupleRowKeySelector[JLong](0),
      new TupleRowKeySelector[JLong](0),
      BasicTypeInfo.LONG_TYPE_INFO,
      1,
      1,
      0)
  }

  // -------------------------------- Matchers --------------------------------

  private def hasFiredCleanUpTimersForTimestamps(timers: JLong*) =
    new TypeSafeMatcher[StubOperatorWithStateTTL]() {

      override protected def matchesSafely(operator: StubOperatorWithStateTTL): Boolean = {
        operator.firedCleanUpTimers.toArray.deep == timers.toArray.deep
      }

      def describeTo(description: Description): Unit = {
        description
          .appendText("a list of timers with timestamps=")
          .appendValue(timers.mkString(","))
      }
    }

  // -------------------------------- Test Classes --------------------------------

  /**
    * A mock [[BaseTwoInputStreamOperatorWithStateRetention]] which registers
    * the timestamps of the clean-up timers that fired (not the registered
    * ones, which can be deleted without firing).
    */
  class StubOperatorWithStateTTL(minRetentionTime: Long, maxRetentionTime: Long)
    extends BaseTwoInputStreamOperatorWithStateRetention(minRetentionTime, maxRetentionTime) {

    val firedCleanUpTimers: mutable.Buffer[JLong] = ArrayBuffer.empty

    override def cleanUpState(time: Long): Unit = {
      firedCleanUpTimers.append(time)
    }

    override def processElement1(element: StreamRecord[CRow]): Unit = {
      registerProcessingCleanUpTimer()
    }

    override def processElement2(element: StreamRecord[CRow]): Unit = {
      registerProcessingCleanUpTimer()
    }

    override def onEventTime(timer: InternalTimer[Any, VoidNamespace]): Unit = ()
  }
}

File: flink-table/flink-table-runtime/src/test/java/org/apache/flink/table/runtime/operators/join/temporal/BaseTwoInputStreamOperatorWithStateRetentionTest.java
/*
 * Licensed to the Apache Software Foundation (ASF) under one
 * or more contributor license agreements.  See the NOTICE file
 * distributed with this work for additional information
 * regarding copyright ownership.  The ASF licenses this file
 * to you under the Apache License, Version 2.0 (the
 * "License"); you may not use this file except in compliance
 * with the License.  You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package org.apache.flink.table.runtime.operators.join.temporal;

import org.apache.flink.api.common.operators.MailboxExecutor;
import org.apache.flink.configuration.CheckpointingOptions;
import org.apache.flink.runtime.execution.Environment;
import org.apache.flink.runtime.state.VoidNamespace;
import org.apache.flink.runtime.state.VoidNamespaceSerializer;
import org.apache.flink.streaming.api.operators.InternalTimer;
import org.apache.flink.streaming.api.operators.InternalTimerService;
import org.apache.flink.streaming.api.operators.SimpleOperatorFactory;
import org.apache.flink.streaming.api.operators.YieldingOperator;
import org.apache.flink.streaming.api.watermark.Watermark;
import org.apache.flink.streaming.runtime.streamrecord.StreamRecord;
import org.apache.flink.streaming.runtime.tasks.StreamTaskMailboxTestHarness;
import org.apache.flink.streaming.runtime.tasks.StreamTaskMailboxTestHarnessBuilder;
import org.apache.flink.streaming.runtime.tasks.TwoInputStreamTask;
import org.apache.flink.table.data.GenericRowData;
import org.apache.flink.table.data.RowData;
import org.apache.flink.table.data.StringData;
import org.apache.flink.table.runtime.keyselector.RowDataKeySelector;
import org.apache.flink.table.runtime.typeutils.InternalTypeInfo;
import org.apache.flink.table.types.logical.BigIntType;
import org.apache.flink.table.types.logical.VarCharType;
import org.apache.flink.table.utils.HandwrittenSelectorUtil;

import org.junit.jupiter.api.Test;

import javax.annotation.Nullable;

import java.time.Duration;
import java.time.Instant;
import java.util.Collections;
import java.util.HashMap;
import java.util.List;
import java.util.Map;
import java.util.stream.Collectors;

import static org.assertj.core.api.Assertions.assertThat;

/**
 * Tests that {@link BaseTwoInputStreamOperatorWithStateRetention} supports interruptible timers
 * with unaligned checkpoints.
 */
class BaseTwoInputStreamOperatorWithStateRetentionTest {

    private static final InternalTypeInfo<RowData> ROW_TYPE =
            InternalTypeInfo.ofFields(new BigIntType(), VarCharType.STRING_TYPE);

    private static final RowDataKeySelector KEY_SELECTOR =
            HandwrittenSelectorUtil.getRowDataSelector(new int[] {0}, ROW_TYPE.toRowFieldTypes());

    @Test
    void testInterruptibleTimersWithWatermarks() throws Exception {
        final Instant firstWindowEnd = Instant.ofEpochMilli(1000L);
        final int numFirstWindowTimers = 2;
        final Instant secondWindowEnd = Instant.ofEpochMilli(2000L);
        final int numSecondWindowTimers = 2;

        try (StreamTaskMailboxTestHarness<RowData> harness =
                createHarness(
                        new StubOperatorWithEventTimeTimers(0, 0)
                                .withTimers(firstWindowEnd, numFirstWindowTimers)
                                .withTimers(secondWindowEnd, numSecondWindowTimers))) {
            harness.processElement(new StreamRecord<>(row(1L, "register")), 0);
            harness.processAll();
            // Advance watermarks on both inputs so the combined watermark progresses.
            harness.processElement(asWatermark(firstWindowEnd), 0);
            harness.processElement(asWatermark(firstWindowEnd), 1);
            harness.processElement(asWatermark(secondWindowEnd), 0);
            harness.processElement(asWatermark(secondWindowEnd), 1);

            List<String> output =
                    harness.getOutput().stream()
                            .map(BaseTwoInputStreamOperatorWithStateRetentionTest::describeRecord)
                            .collect(Collectors.toList());
            assertThat(output)
                    .containsExactly(
                            firedDesc(0L),
                            mailDesc(0L),
                            firedDesc(1L),
                            mailDesc(1L),
                            watermarkDesc(firstWindowEnd),
                            firedDesc(0L),
                            mailDesc(0L),
                            firedDesc(1L),
                            mailDesc(1L),
                            watermarkDesc(secondWindowEnd));
        }
    }

    private StreamTaskMailboxTestHarness<RowData> createHarness(
            StubOperatorWithEventTimeTimers operator) throws Exception {
        return new StreamTaskMailboxTestHarnessBuilder<>(
                        (Environment env) -> new TwoInputStreamTask<>(env), ROW_TYPE)
                .addJobConfig(CheckpointingOptions.CHECKPOINTING_INTERVAL, Duration.ofSeconds(1))
                .addJobConfig(CheckpointingOptions.ENABLE_UNALIGNED, true)
                .addJobConfig(CheckpointingOptions.ENABLE_UNALIGNED_INTERRUPTIBLE_TIMERS, true)
                .setKeyType(KEY_SELECTOR.getProducedType())
                .addInput(ROW_TYPE, 1, KEY_SELECTOR)
                .addInput(ROW_TYPE, 1, KEY_SELECTOR)
                .setupOperatorChain(SimpleOperatorFactory.of(operator))
                .name("temporal-join")
                .finishForSingletonOperatorChain(ROW_TYPE.createSerializer(null))
                .build();
    }

    private static RowData row(long key, String value) {
        return GenericRowData.of(key, StringData.fromString(value));
    }

    private static Watermark asWatermark(Instant timestamp) {
        return new Watermark(timestamp.toEpochMilli());
    }

    private static String describeRecord(Object obj) {
        if (obj instanceof Watermark) {
            return "Watermark@" + ((Watermark) obj).getTimestamp();
        } else if (obj instanceof StreamRecord) {
            Object value = ((StreamRecord<?>) obj).getValue();
            if (value instanceof RowData) {
                RowData row = (RowData) value;
                return row.getLong(0) + "," + row.getString(1).toString();
            }
            return value.toString();
        }
        return obj.toString();
    }

    private static String firedDesc(long key) {
        return key + ",fired";
    }

    private static String mailDesc(long key) {
        return key + ",mail";
    }

    private static String watermarkDesc(Instant timestamp) {
        return "Watermark@" + timestamp.toEpochMilli();
    }

    /**
     * A stub {@link BaseTwoInputStreamOperatorWithStateRetention} that registers event-time timers
     * and emits records when they fire, used to verify interruptible timer support.
     */
    private static class StubOperatorWithEventTimeTimers
            extends BaseTwoInputStreamOperatorWithStateRetention
            implements YieldingOperator<RowData> {

        private final Map<Instant, Integer> timersToRegister;
        private transient @Nullable MailboxExecutor mailboxExecutor;
        private transient InternalTimerService<VoidNamespace> eventTimeTimerService;

        StubOperatorWithEventTimeTimers(long minRetentionTime, long maxRetentionTime) {
            this(minRetentionTime, maxRetentionTime, Collections.emptyMap());
        }

        StubOperatorWithEventTimeTimers(
                long minRetentionTime,
                long maxRetentionTime,
                Map<Instant, Integer> timersToRegister) {
            super(minRetentionTime, maxRetentionTime);
            this.timersToRegister = timersToRegister;
        }

        @Override
        public void setMailboxExecutor(MailboxExecutor mailboxExecutor) {
            super.setMailboxExecutor(mailboxExecutor);
            this.mailboxExecutor = mailboxExecutor;
        }

        @Override
        public void open() throws Exception {
            super.open();
            eventTimeTimerService =
                    getInternalTimerService("event-timers", VoidNamespaceSerializer.INSTANCE, this);
        }

        @Override
        public void processElement1(StreamRecord<RowData> element) {
            registerTimers();
        }

        @Override
        public void processElement2(StreamRecord<RowData> element) {
            registerTimers();
        }

        private void registerTimers() {
            for (Map.Entry<Instant, Integer> entry : timersToRegister.entrySet()) {
                for (int keyIdx = 0; keyIdx < entry.getValue(); keyIdx++) {
                    setCurrentKey(GenericRowData.of((long) keyIdx, null));
                    eventTimeTimerService.registerEventTimeTimer(
                            VoidNamespace.INSTANCE, entry.getKey().toEpochMilli());
                }
            }
        }

        @Override
        public void onEventTime(InternalTimer<Object, VoidNamespace> timer) throws Exception {
            RowData key = (RowData) timer.getKey();
            long keyValue = key.getLong(0);
            mailboxExecutor.execute(
                    () -> output.collect(new StreamRecord<>(row(keyValue, "mail"))),
                    "mail-" + keyValue);
            output.collect(new StreamRecord<>(row(keyValue, "fired")));
        }

        @Override
        public void cleanupState(long time) {}

        StubOperatorWithEventTimeTimers withTimers(Instant timestamp, int count) {
            Map<Instant, Integer> copy = new HashMap<>(timersToRegister);
            copy.put(timestamp, count);
            return new StubOperatorWithEventTimeTimers(0, 0, copy);
        }
    }
}
